src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,33 +852,48 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
                 return 0;
             }
 
-            fixed (char* pSource = source)
-            {
-                int sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, null, 0, options);
+            // according to ICU User Guide the performance of ucol_getSortKey is worse when it is called with null output buffer
+            // the solution is to try to fill the sort key in a temporary buffer of size equal 4 x string length
+            // 1MB is the biggest array that can be rented from ArrayPool.Shared without memory allocation
+            int sortKeyLength = (source.Length > 1024 * 1024 / 4) ? 0 : 4 * source.Length;
 
-                byte[]? borrowedArr = null;
-                Span<byte> span = sortKeyLength <= 512 ?
-                    stackalloc byte[512] :
-                    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
+            byte[]? borrowedArray = null;
+            Span<byte> sortKey = sortKeyLength <= 512
+                ? stackalloc byte[512]
+                : (borrowedArray = ArrayPool<byte>.Shared.Rent(sortKeyLength));
 
-                fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
+            fixed (char* pSource = &MemoryMarshal.GetReference(source))
```

Comment:
We seem to commonly combine nested usings/fixed that have the same scope, if you want.
```c#
            fixed (char* pSource = &MemoryMarshal.GetReference(source))
            fixed (byte* pSortKey = &MemoryMarshal.GetReference(sortKey))
```

Reply:
the value of `sortKey` changes in the slow path scenario for big strings, so I could not pin both buffers here